pkg/analysis/nomaps/analyzer.go

```diff
@@ -0,0 +1,89 @@
+package nomaps
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/types"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const name = "nomaps"
+
+var (
+	errCouldNotGetInspector = errors.New("could not get inspector")
+)
+
+type analyzer struct{}
+
+// newAnalyzer creates a new analyzer.
+func newAnalyzer() *analysis.Analyzer {
+	a := &analyzer{}
+
+	return &analysis.Analyzer{
+		Name:     name,
+		Doc:      "no map type should be allowed and prefer list type",
+		Run:      a.run,
+		Requires: []*analysis.Analyzer{inspect.Analyzer},
+	}
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (interface{}, error) {
+	inspect, ok := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	if !ok {
+		return nil, errCouldNotGetInspector
+	}
+
+	// Filter to structs so that we can iterate over fields in a struct.
+	nodeFilter := []ast.Node{
+		(*ast.StructType)(nil),
+	}
```

Comment:
A recent refactor introduced a shared iterator for checking fields that provides additional information and checks for fields that we shouldn't be looking at (those in function declarations, those with json `-` ignoring them).

Do you think we should leverage that here too now?

https://github.com/JoelSpeed/kal/blob/abd233a9fed84968640bcfdb40f375549047fb07/pkg/analysis/commentstart/analyzer.go#L33-L40

Alternatively, does the type checker work here?

https://github.com/JoelSpeed/kal/blob/abd233a9fed84968640bcfdb40f375549047fb07/pkg/analysis/nofloats/analyzer.go#L41-L50

Or perhaps that deconstructs maps already to the point where it's not useful for us

Reply:
type checker checks type recursively, so it might be not useful. 
But shared iterator is what we should use. Thanks!

Reply:
If we want to check type declaration too, we can't use shared iterator.

Reply:
Lets ignore the type declarations for now then and see how we get on, you're right on the other thread, we are capturing what's used at the field level already